### PR TITLE
Wip/transformers bump 3.5.1

### PR DESCRIPTION
**Pull request by Mehrad0711**
This PR bumps transformers version to 3.5.1.
Seq2Seq models now return cross-attention weights which allow substantial code refactoring and cleanup in `transformers_utils`.
Model outputs are now NamedTuples (e.g. Seq2SeqLMOutput) which allows key indexing as well as slicing.

P.S. Maybe it's time to push a new pip version!

### CODE REVIEW

**Comment by s-jse** [comment]
Since everything else (our VMs and the Pipfile here) have been upgraded to Python 3.8, should we change the Travis build configuration to also use Python 3.8, @gcampax ?

**Comment by gcampax** [comment]
Ideally, we would test multiple versions of Python, because we don't want to force people who are looking to use genienlp to the very latest Python. But practically, we would run out of credits very quickly if we did that. So leaving the oldest version of Python we still support (3.6) is probably best.

**Comment by gcampax** [comment]
It's also worth mentioning that we don't really use pipenv anywhere outside of Travis itself, so maybe the answer is to remove pipenv entirely...

File: setup.py [hunk 1]
```diff
@@ -56,8 +56,8 @@
         'pyrouge>=0.1.3',
         'sacrebleu~=1.0',
         'requests~=2.22',
-        'transformers==2.11',
-        'sentencepiece>=0.1.83,<0.2.0',
+        'transformers==3.5.1',
+        'sentencepiece==0.1.91',
```
Comment (gcampax) [inline setup.py]: Do we really need these specific versions, or can we relax the requirements a bit?

File: genienlp/paraphrase/transformers_utils.py [hunk 2]
```diff
@@ -63,1096 +84,257 @@
...
-            "decoder_input_ids": decoder_input_ids,
-            "attention_mask": attention_mask,
-            "use_cache": use_cache,  # change this to avoid caching (presumably for debugging)
-        }
 
-    def _generate_no_beam_search(
+class GeniePreTrainedModel(PreTrainedModel):
+    '''
+    General class for PreTrainedModel which can output cross-attention weights during generation
+    '''
```
Comment (gcampax) [inline genienlp/paraphrase/transformers_utils.py:122]: I thought cross-attention weights were natively supported now?

**Review by gcampax — COMMENTED** [review]
Generally speaking this makes sense. A couple comments. Also, I'm looking forward to more cleanups in the future.

File: setup.py (hunk 1 above)
Comment (Mehrad0711) [reply setup.py]: huggingface pins sentencepiece to 0.1.91.  Not sure if it works with other versions...

File: genienlp/paraphrase/transformers_utils.py (hunk 2 above)
Comment (Mehrad0711) [reply genienlp/paraphrase/transformers_utils.py:122]: The models output cross-attention in their `forward()` method, but running `generation()` still doesn't return them. Maybe we should make a PR for this.

File: setup.py (hunk 1 above)
Comment (gcampax) [reply setup.py]: But we don't need that specific version, so you should leave our requirement more open and then pip will pick up the more stringent requirement from huggingface.

File: genienlp/paraphrase/transformers_utils.py (hunk 2 above)
Comment (gcampax) [reply genienlp/paraphrase/transformers_utils.py:122]: Please do! Let's get rid of this code...

Comment (Mehrad0711) [reply genienlp/paraphrase/transformers_utils.py:122]: One thing though is that we need a subclass like this for returning other useful stuff such as confidence scores that `generation()` doesn't return. @s-jse may want to speak of his experience so far :)

Comment (s-jse) [reply genienlp/paraphrase/transformers_utils.py:122]: Yeah, we need a `GeniePreTrainedModel` class, we could get rid of the rest I think (unless a reason for model-specific behavior comes up in the future). Specifically, I need the generation code to output logits, and potentially other useful things that help us determine the model confidence.

**Review by s-jse — APPROVED** [review]
Looks good. I tested and the BART models we trained previously run with this version as well.

**Comment by Mehrad0711** [comment]
Seems like Travis doesn't like running again. I ran the tests locally.
